pallets/transfer-allowlist/src/benchmarking.rs

```diff
@@ -0,0 +1,298 @@
+use cfg_traits::fees::Fees;
+use cfg_types::{locations::Location, tokens::CurrencyId};
+use codec::EncodeLike;
+use frame_benchmarking::*;
+use frame_support::traits::{Currency, Get, ReservableCurrency};
+use frame_system::RawOrigin;
+use scale_info::TypeInfo;
+use sp_runtime::traits::{AtLeast32BitUnsigned, Bounded, CheckedAdd, One};
+
+use super::*;
+benchmarks! {
+	where_clause {
+	  where
+		T: Config<CurrencyId = CurrencyId, Location = Location>,
+	<T as frame_system::Config>::AccountId: Into<<T as pallet::Config>::Location>,
+	  <T as pallet::Config>::Location: From<<T as frame_system::Config>::AccountId> + EncodeLike<<T as pallet::Config>::Location>,
+	  <T as pallet::Config>::ReserveCurrency: Currency<<T as frame_system::Config>::AccountId> + ReservableCurrency<<T as frame_system::Config>::AccountId>,
+	  <T as frame_system::Config>::BlockNumber: AtLeast32BitUnsigned + Bounded + TypeInfo
+
+}
+
+	add_transfer_allowance_no_existing_metadata {
+	let (sender, receiver) = set_up_users::<T>();
+	}:add_transfer_allowance(RawOrigin::Signed(sender.clone()), CurrencyId::Native, receiver.clone().into())
+	verify {
+	  assert_eq!(
+				Pallet::<T>::get_account_currency_transfer_allowance(
+					  (sender,
+					  CurrencyId::Native,
+					  Location::from(receiver))
+				).unwrap(),
+				AllowanceDetails {
+					  allowed_at: T::BlockNumber::zero(),
+					  blocked_at: T::BlockNumber::max_value(),
+				}
+		  )
+
+  }
+
+	  add_transfer_allowance_existing_metadata {
+		  let (sender, receiver) = set_up_users::<T>();
+		  Pallet::<T>::add_allowance_delay(RawOrigin::Signed(sender.clone()).into(), CurrencyId::Native, 200u32.into())?;
+	  }:add_transfer_allowance(RawOrigin::Signed(sender.clone()), CurrencyId::Native, receiver.clone().into())
+		  verify {
+			  assert_eq!(
+						Pallet::<T>::get_account_currency_transfer_allowance(
+							  (sender,
+							   CurrencyId::Native,
+							   Location::from(receiver))
+						).unwrap(),
+						AllowanceDetails {
+							  allowed_at: <frame_system::Pallet<T>>::block_number() + 200u32.into(),
+							  blocked_at: T::BlockNumber::max_value(),
+						}
+				)
+
+		}
+
+  add_allowance_delay_no_existing_metadata {
+		let (sender, receiver) = set_up_users::<T>();
+  }:add_allowance_delay(RawOrigin::Signed(sender.clone()), CurrencyId::Native, 200u32.into())
+  verify {
+	  assert_eq!(
+		  Pallet::<T>::get_account_currency_restriction_count_delay(
+			  sender,
+						CurrencyId::Native,
+		  ).unwrap(),
+		  AllowanceMetadata {
+			  allowance_count: 0,
+			  current_delay: Some(200u32.into()),
+			  once_modifiable_after: None
+		  }
+	  )
+  }
+	add_allowance_delay_existing_metadata {
+			let (sender, receiver) = set_up_users::<T>();
+			Pallet::<T>::add_transfer_allowance(RawOrigin::Signed(sender.clone()).into(), CurrencyId::Native, receiver.clone().into())?;
+	}:add_allowance_delay(RawOrigin::Signed(sender.clone()), CurrencyId::Native, 200u32.into())
```

Comment:
I see you're benchmarking the same extrinsic with different setups, I think this is not needed. Choosing the worst case is enough. (This applies to other extrinsics)

Reply:
In this particular case I wanted to make sure that the worst case was covered in case creating the metadata becomes more expensive in the future -- right now it's faster as it's writing the same amount when creating, but just pulling from default values.  I can remove the no existing metadata benchmark though.

Reply:
I disagree with @lemunozm. Benchmarks, if done well and right, make the chain more efficient. Albeit knowing, that reads and writes are probably at all times the heavier instructions, we should not neglect the fact that we should bench all codepaths and let the system decide what is the heaviest. 

If we do the weights correctly, we should return the overweight fees via `DispatchWithPostInfo`. But defaulting to the above is not a good idea IMO.

Reply:
What you say is the next step. I mean, the first step is to determine and make a benchmark for the worst case (the safety step). Later the second step is what you say, where you fix your extremely bad measured extrinsic with `DispatchWithPostInfo` for the no-worst cases (the performance step).

Reply:
Anyways I think these performance improvements should be evaluated to see if it worth it to do it or not, and if we are at the point to do it. Probably in simple functions like these ones, we do no need to optimize for this. And doing it adds a lot of complexity for very little improvement. As the famous sentence says: “Premature optimization is the root of all evil”

Reply:
@mustermeiszer Looking the implementation of both paths, they are almost the same. Not sure if we want to add a `max()` complexity for an improvement close to the benchmark noise.

Reply:
I am not strongly against removing it in this case. It is just not that big of "not-good-looking-style" or complexity to have a single max call at the head of an extrinsic definition. But it ensures we are not missing possible changes when the extrinsic gets heavier in the future. 

I would always lean for more security.

Reply:
> to have a single max call at the head of an extrinsic definition

It's not only that, for me, the big issue is to maintain the double benchmarks

Of course, each case has its own complexity and should be evaluated if it needs extra benchmarks for its different internal paths. Maybe I've sounded very strong in my first message against this, there are cases that need what you say.

Reply:
I'm thinking atm that it's simpler and safer to have the multiple benchmarks for the different scenarios, than to place the burden on any future work that might need to be done with the pallet. I feel that without having to have the verify in the benchmarks maintaining the benchmarks for different scenarios is comparatively simple.